### PR TITLE
Browser Stat/Verb Panel

### PR DESCRIPTION
**Pull request by ItsSelis**
This PR is in a functional, yet experimental, state. The verbs still all work fine and nicely (with the MC tab for example even having better performance/no lag). Though some visual bugs could occur.
Please report such bugs in the comments or make a new isssue.

@Cameron653 has made a simple preview of the current state (as of commit [41c0470](https://github.com/VOREStation/VOREStation/pull/13816/commits/41c047008d573dfc84cd643efd35cfac178bd8a9)):

The stuff list:
- [x] (!) Fix categories disappear (but verbs stay) when joining the round (on the admin tab)
- [x] ( ) Categorizing the verbs (or some of them at least, if wanted)
- [ ] ( ) Pref for either splitting sub-categories of admin tab into tabs or leave it categorized under admin (Default: No split)
- [x] ( ) Removing the Interviews stuff (tickets stay of course)
- [x] (!) Globals/Master-/Failsafe-controller not showing any stats (but runtimes are fixed)
- [x] (!!) Implement switching to dark mode (necessary styles and functions do exist already)
- [x] (!!!) Make Examine panel work
- [x] (!!!) Less Icon lag (Turf examination)
- [ ] desc_panel_image should use icon2html instead of the old code to properly render interaction icons
- [ ] Loading the server with me as a host does not give me access to the tools in the admin panel until I load in as an observer. I otherwise only have what you see in the image below.
- [x] Observe button (at least as a host role) does not make you observe and instead just removes the menu, apparently spawns a ghost of your current slot, and leaves you on the main screen. Attempting to join the round normally does work, however.
- [ ] You cannot pick things up from the turf examine menu by clicking on them.

### CODE REVIEW

File: code/controllers/subsystems/statpanel.dm [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	init_order = INIT_ORDER_STATPANELS
+	runlevels = RUNLEVELS_DEFAULT | RUNLEVEL_LOBBY
+	var/list/currentrun = list()
+	var/encoded_global_data
+	var/mc_data_encoded
+	var/list/cached_images = list()
+
+/datum/controller/subsystem/statpanels/fire(resumed = FALSE)
+	if (!resumed)
+		var/datum/map_config/cached = SSmapping.next_map_config
```
Comment (Cameron653) [inline code/controllers/subsystems/statpanel.dm:13]: Map config related things can be found here https://github.com/tgstation/tgstation/blob/17e6dc38d7da5e16a62c28fbb5313f66a295ade6/code/datums/map_config.dm

(Errors on line 13, 14, 16, 17, 18)

File: code/controllers/subsystems/statpanel.dm [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+			"Map: [SSmapping.config?.map_name || "Loading..."]",
+			cached ? "Next Map: [cached.map_name]" : null,
+			"Round ID: [GLOB.round_id ? GLOB.round_id : "NULL"]",
+			"Server Time: [time2text(world.timeofday, "YYYY-MM-DD hh:mm:ss")]",
+			"Round Time: [round_time > MIDNIGHT_ROLLOVER ? "[round(round_time/MIDNIGHT_ROLLOVER)]:[worldtime2stationtime()]" : worldtime2stationtime()]",
+			"Station Time: [stationtime2text()]",
+			"Time Dilation: [round(SStime_track.time_dilation_current,1)]% AVG:([round(SStime_track.time_dilation_avg_fast,1)]%, [round(SStime_track.time_dilation_avg,1)]%, [round(SStime_track.time_dilation_avg_slow,1)]%)"
+		)
+
+		if(SSshuttle.emergency)
```
Comment (Cameron653) [inline code/controllers/subsystems/statpanel.dm:25]: This line more than likely needs to be changed to:
```c
if(emergency_shuttle_controller.emergency_shuttle.evac)
```

Our shuttle subsystem is here. https://github.com/VOREStation/VOREStation/blob/60011ecd0b047d789fb7fdcdd7336dde4aa8a584/code/controllers/subsystems/shuttles.dm

And more specifically, our emergency shuttle controller is here. https://github.com/VOREStation/VOREStation/blob/5195f715452d7a92f2b911f8217343a856f6ed13/code/controllers/emergency_shuttle_controller.dm

File: code/controllers/subsystems/statpanel.dm [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+			var/ETA = SSshuttle.emergency.getModeStr()
```
Comment (Cameron653) [inline code/controllers/subsystems/statpanel.dm:26]: Most likely going to be need to be changed to
```c
var/ETA = emergency_shuttle_controller.emergency_shuttle.get_status_panel_eta()
```

See: https://github.com/VOREStation/VOREStation/blob/5195f715452d7a92f2b911f8217343a856f6ed13/code/controllers/emergency_shuttle_controller.dm#L227

File: code/controllers/subsystems/statpanel.dm [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+			if(ETA)
+				global_data += "[ETA] [SSshuttle.emergency.getTimerStr()]"
```
Comment (Cameron653) [inline code/controllers/subsystems/statpanel.dm:28]: Should be able to change it to just 
```c
global_data += "[ETA]
```

**Review by Cameron653 — COMMENTED** [review]
Will note more things down when I get home.

File: code/controllers/subsystems/statpanel.dm (hunk 2 above)
Comment (ItsSelis) [reply code/controllers/subsystems/statpanel.dm:25]: Yeah, it worked with "emergency_shuttle.evac".
emergency_shuttle seems to be already a global var of the controller itself (emergency_shuttle_controller) or something.

**Comment by Cameron653** [comment]
The current rough draft:

File: html/statbrowser.js [hunk 5]
```diff
@@ -0,0 +1,1003 @@
...
+	if (sdql2) {
+		sdql2 = [];
+		removePermanentTab("SDQL2");
+		if (current_tab == "SDQL2")
+			tab_change("Status");
+	}
+	checkStatusTab();
+}
+
+function remove_interviews() {
```
Comment (Cameron653) [inline html/statbrowser.js:416]: Don't have interviews.

File: html/statbrowser.js [hunk 6]
```diff
@@ -0,0 +1,1003 @@
...
+	addPermanentTab(turfname);
+	tab_change(turfname);
+});
+
+Byond.subscribeTo('remove_admin_tabs', function () {
+	href_token = null;
+	remove_mc();
+	remove_tickets();
+	remove_sdql2();
+	remove_interviews();
```
Comment (Cameron653) [inline html/statbrowser.js:938]: More interview stuff.

**Review by Cameron653 — COMMENTED** [review]
Removal of the interview mentions would be good and also a small change that tells people how to fix any issues that may arise in case an admin deletes their stat_panel var somehow.

File: code/modules/admin/topic.dm [hunk 7]
```diff
@@ -1,3 +1,19 @@
+/datum/admins/proc/CheckAdminHref(href, href_list)
+	var/auth = href_list["admin_token"]
+	. = auth && (auth == href_token || auth == GLOB.href_token)
+	if(.)
+		return
+	var/msg = !auth ? "no" : "a bad"
+	message_admins("[key_name_admin(usr)] clicked an href with [msg] authorization key!")
+
+	var/debug_admin_hrefs = TRUE // Remove once everything is converted over
```
Comment (ItsSelis) [inline code/modules/admin/topic.dm:9]: - [ ] Remove once PR is ready

**Comment by Cameron653** [comment]
Work in progress, but just an example of it reverting back to the old style if TGUI ever does break.


And what the panel looks like at the moment. Still WIP.
Discord_2022-10-01_16-39-08

**Comment by Heroman3003** [comment]
The only concern potential I have towards this is whether there is generally a proper fallback back to default BYOND stat/verb panels; as well as whether it could potentially be just client-toggled in case its found to cause performance issues clientside.

**Comment by ItsSelis** [comment]
> The only concern potential I have towards this is whether there is generally a proper fallback back to default BYOND stat/verb panels; as well as whether it could potentially be just client-toggled in case its found to cause performance issues clientside.

It is theoretically, maybe, possible to implement such. Though it is only little feasible as it was tested even with Internet Explorer 8 with it still working fine. Thpugh this also still is a draft which will get things added or changed up anyways.

**Comment by Heroman3003** [comment]
> It is theoretically, maybe, possible to implement such. Though it is only little feasible as it was tested even with Internet Explorer 8 with it still working fine. Thpugh this also still is a draft which will get things added or changed up anyways.

I am not saying it should have it RIGHT NOW in current state or anything. Just that ideally when its complete, it having a toggle to switch back to byond default and using byond default as a fallback for any 'it broke' level issues automatically (akin to vchat) would be good.

**Comment by ItsSelis** [comment]
> > It is theoretically, maybe, possible to implement such. Though it is only little feasible as it was tested even with Internet Explorer 8 with it still working fine. Thpugh this also still is a draft which will get things added or changed up anyways.
> 
> I am not saying it should have it RIGHT NOW in current state or anything. Just that ideally when its complete, it having a toggle to switch back to byond default and using byond default as a fallback for any 'it broke' level issues automatically (akin to vchat) would be good.

Of course, of course. Cameron already suggested such earlier already, so it was always in the back of our heads. Just wanted to get it running first xP

**Comment by ItsSelis** [comment]
Still open to be fixed:
- [ ] Pref for either splitting sub-categories of admin tab into tabs or leave it categorized under admin (current default is set to **NOT** split the verbs) 
- [ ] desc_panel_image should use icon2html instead of the old code to properly render interaction icons

**Comment by TankTheBirb** [comment]
Issues found after testing:
- Loading the server with me as a host does not give me access to the tools in the admin panel until I load in as an observer. I otherwise only have what you see in the image below.
image
- Observe button (at least as a host role) does not make you observe and instead just removes the menu, apparently spawns a ghost of your current slot, and leaves you on the main screen. Attempting to join the round normally does work, however.
- You cannot pick things up from the turf examine menu by clicking on them.

No other issues found from my brief examination of the PR. Notes:
- Sorted  menus and categories in windows is really fuckin good and it's to the point where we could reasonably reduce the amount of admin tabs further (recommending for a later PR to avoid scope creep)
- Accessing the MC is basically lagfree currently. I encountered no issues opening debug menus and playing around with it
- Turf examine is INCREDIBLY fast as it took less than half a second to turf examine a tile of 500 signalers. Tho I did have a random crash when I walked over them, which is probably unrelated given I could not reproduce the issue.

In short I think this PR has promise and is near completion but probably needs a bit of tuning and bit of testing to be assumed green across the board and ready for merge. That's all I can do for now, @ItsSelis Please let me know on discord if you update this PR further and I will try my best to get to doing further testing.

**Comment by Heroman3003** [comment]
Alright. With issues brought up and to-do list still incomplete, I'll be waiting until all those are handled before proceeding with merging of it.

**Comment by ItsSelis** [comment]
This will be re-done with a clean start and proper implementation in a new PR!
